lib/Dialect/MSFT/MSFTPasses.cpp

```diff
@@ -319,34 +325,189 @@ void PartitionPass::runOnOperation() {
+         hw::isAnyModule(module);
+}
+hw::ModulePortInfo getModulePortInfo(Operation *op) {
+  if (auto mod = dyn_cast<MSFTModuleOp>(op))
+    return mod.getPorts();
+  if (auto mod = dyn_cast<MSFTModuleExternOp>(op))
+    return mod.getPorts();
+  return hw::getModulePortInfo(op);
+}
+
+/// Heuristics to get the entity name.
+static StringRef getOpName(Operation *op) {
+  StringAttr name;
+  if ((name = op->getAttrOfType<StringAttr>("name")) && name.size())
+    return name.getValue();
+  if ((name = op->getAttrOfType<StringAttr>("sym_name")) && name.size())
+    return name.getValue();
+  return op->getName().getStringRef();
+}
+/// Try to set the entity name.
+/// TODO: this needs to be more complex to deal with renaming symbols.
+static void setEntityName(Operation *op, Twine name) {
+  StringAttr nameAttr = StringAttr::get(op->getContext(), name);
+  if (op->hasAttrOfType<StringAttr>("name"))
+    op->setAttr("name", nameAttr);
+  if (op->hasAttrOfType<StringAttr>("sym_name"))
+    op->setAttr("sym_name", nameAttr);
+}
+/// Heuristics to get the output name.
+static StringRef getResultName(OpResult res, const hw::SymbolCache &syms,
+                               std::string &buff) {
+  Operation *op = res.getDefiningOp();
+  if (auto inst = dyn_cast<InstanceOp>(op)) {
+    Operation *modOp = syms.getDefinition(inst.moduleNameAttr());
+    assert(modOp && "Invalid IR");
+    assert(isAnyModule(modOp) && "Instance must point to a module");
+    hw::ModulePortInfo ports = getModulePortInfo(modOp);
+    return ports.outputs[res.getResultNumber()].name;
+  }
+  if (auto asmInterface = dyn_cast<mlir::OpAsmOpInterface>(op)) {
+    StringRef retName;
+    asmInterface.getAsmResultNames([&](Value v, StringRef name) {
+      if (v == res && !name.empty())
+        retName = StringAttr::get(op->getContext(), name).getValue();
+    });
+    if (!retName.empty())
+      return retName;
+  }
 
-  // TODO: For operations which target partitions not in the same module, bubble
-  // them up.
+  // Fallback. Not ideal.
+  buff.clear();
+  llvm::raw_string_ostream(buff) << "out" << res.getResultNumber();
+  return buff;
+}
+
+/// Heuristics to get the input name.
+static StringRef getOperandName(OpOperand &oper, const hw::SymbolCache &syms,
+                                std::string &buff) {
+  Operation *op = oper.getOwner();
+  if (auto inst = dyn_cast<InstanceOp>(op)) {
+    Operation *modOp = syms.getDefinition(inst.moduleNameAttr());
+    assert(modOp && "Invalid IR");
+    assert(isAnyModule(modOp) && "Instance must point to a module");
+    hw::ModulePortInfo ports = getModulePortInfo(modOp);
+    return ports.inputs[oper.getOperandNumber()].name;
+  }
+
+  // Fallback. Not ideal.
+  buff.clear();
+  llvm::raw_string_ostream(buff) << "in" << oper.getOperandNumber();
+  return buff;
+}
+
+void PartitionPass::bubbleUp(MSFTModuleOp mod, ArrayRef<Operation *> ops) {
+  // This particular implementation is _very_ sensitive to iteration order. It
```

Comment:
I suspect this'll change when I get around to floating wires/ports.

Reply:
Makes sense, this simplification seems to lead to a very clear implementation right now.